content/backend/graphql-scala/2-preparing-first-query.md

```diff
@@ -0,0 +1,377 @@
+* **9** `Executor.execute` is the most important call in this class because it's the point where query is executed. If executor responds with success, the result is send back to the client, in the other case server will responds with status code 4xx and some kind of explanation what was wrong with the query. Executor needs some data to fulfil the request. Three of them are `query`(**11**), `operationName`(**13**) and `variables`(**14**) which are read from request.
+The last two are: `GraphQLSchema.SchemaDefinition` and `MyContext(dao)`.
+* **12** MyContext is a context object mentioned above. In this case we're building the context object with prepared DAO class.
+* `GraphQLSchema.SchemaDefinition` is the last object we have to explain here. It contains our Schema - what we are able to query for. It also interprets how data is fetched and from which data source (i.e. one or more databases, REST call to the other server…). In short our `SchemaDefinition` file defines what we want to expose. There are defined types (from GraphQL point of view) and shape of the schema a client is able to query for.
+
+### Define GraphQLSchema
+
+<Instruction>
+
+Create GraphQLSchema object:
+
+```scala
+import com.howtographql.scala.sangria.models.Link
+import sangria.schema.{Field, ListType, ObjectType}
+// #
+import sangria.schema._
+import sangria.macros.derive._
+
+object GraphQLSchema {
+
+  // 1
+  implicit val LinkType = deriveObjectType[Unit, Link]()
+
+  // 2
+  val QueryType = ObjectType(
+    "Query",
+    fields[MyContext, Unit](
+      Field("allLinks", ListType(LinkType), resolve = c => c.ctx.dao.allLinks)
+    )
+  )
+
+  // 3
+  val SchemaDefinition = Schema(QueryType)
+}
+```
+
+</Instruction>
+
+Sangria cannot reuse case classes defied in our domain, it needs it's own object called of type `ObjectType`. On the other hand it no need to be 1-1 mapping between our Sangria and DB models. This abstraction allows us to freely hide, add or aggregate fields.
+* **1** is a definition of ObjectType for our `Link` class. `deriveObjectType` is a macro what will built an Object for us. But we can do it manually. For example `deriveObjectType[Unit, Link]()` is the same to
+
+```scala
+val LinkType = ObjectType[Unit, Link](
+     "Link",
+     fields[Unit, Link](
+       Field("id", IntType, resolve = _.value.id),
+       Field("url", StringType, resolve = _.value.url),
+       Field("description", StringType, resolve = _.value.description)
+     )
+   )
+```
+
+When you want to use macro-way to define objects don't forget to import `sangria.macros.derive._`
+
+* **2** `val QueryType` is a top level object of our schema. Probably it also could be defined by macro but I decided to make it manually. As you can see, the top level object has name `Query` and it (along with nested object) will be available to see in the graphiql console what we will include farther in this chapter. In `fields` definition I've added only one `Field` at this moment.
```

Comment:
Did you introduce the short then log version on purpose? Often you see tutorials first giving the "long" way and explaining it, then showing a "magic" to avoid boilerplate. I wonder why you went the other way. Is the manual mapping not useful in real life?